### PR TITLE
feat(interchain-token-service): add interchain_transfer implementation

### PR DESCRIPTION
**Pull request by milapsheth**
[AXE-6744]

[AXE-6744]: https://axelarnetwork.atlassian.net/browse/AXE-6744?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/interchain-token-service/tests/interchain_transfer.rs [hunk 1]
```diff
@@ -109,7 +111,9 @@ fn interchain_transfer_receive() {
 
     gateway_client.approve_messages(&messages, &proof);
 
-    client.execute(&source_chain, &message_id, &source_address, &payload);
+    client
```
[the remark this answers is not in the document]
Comment (milapsheth) [reply contracts/interchain-token-service/tests/interchain_transfer.rs:114]: yes, I was discussing with Laith this week in the gateway tests PR. We should avoid bypassing auth checks at the `env` so we don't miss some authentication by mistake. We can mock it for specific invocations when we know it's not needed, but by default we should authenticate properly. Will put my thoughts in slack later for a consistent approach

File: contracts/interchain-token-service/tests/utils/mod.rs [hunk 2]
```diff
@@ -26,6 +28,9 @@ fn setup_its<'a>(
     let owner = Address::generate(env);
     let its_hub_address = String::from_str(env, "its_hub_address");
     let chain_name = String::from_str(env, "chain_name");
+
+    // Note: On changes to `interchain-token` crate, recompile it via `stellar contract build && ./optimize.sh`
```
[the remark this answers is not in the document]
Comment (milapsheth) [reply contracts/interchain-token-service/tests/utils/mod.rs:32]: As a pattern, I recommend optimizing just so we don't run into bytecode size limit issues if the unoptimized version exceeds the size

File: contracts/interchain-token-service/src/contract.rs [hunk 3]
```diff
@@ -257,19 +258,33 @@ impl InterchainTokenServiceInterface for InterchainTokenService {
...
         amount: i128,
-        metadata: Option<Bytes>,
+        data: Option<Bytes>,
         gas_token: Token,
     ) -> Result<(), ContractError> {
-        caller.require_auth();
+        ensure!(amount > 0, ContractError::InvalidAmount);
 
-        // TODO: implementation
+        token_handler::take_token(
```
[the remark this answers is not in the document]
Comment (milapsheth) [reply contracts/interchain-token-service/src/contract.rs:266]: That EVM logic doesn't exist in Stellar atm, so not needed

File: contracts/interchain-token-service/tests/utils/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahramy) [reply contracts/interchain-token-service/tests/utils/mod.rs:32]: I see. I agree with that, we should deploy only optimized contracts.

**Review by ahramy — APPROVED** [review]
LGTM!

**Pull request merged**